### PR TITLE
[Feature] Litecoin card v1

### PR DESCRIPTION
**Pull request by kcw-grunt**
## Overview
This is work has been in progress for over 1 year and we are now ready to release it as an **ALPHA** to our users.  This feature allows users to Register and Login to the Litecoin Card (https://litecoin-card.com/)

The next step is to add transfer and syncing capabilities

## Approach
The demo of this work used a combination of UIKit versions with programmitic and Storyboard layouts.  The blockcard was not yet  completely available.  To allow better visualizations and quicker developments (as the risk of an incomplete framework) SwiftUI and MVVM is used primarily in this work.

## Components
Some parts of the work can be described in these areas:
- Hybrid UIKit / SwiftUI integrations
- Legacy mockups used as a guide but will be polished further 
- Using the Partner Library API

## Clips

| Before | After|
|--------|-----|
|v2 8 30-fix|CardV2|

### CODE REVIEW

File: loafwallet/src/Strings/Base.lproj/Localizable.strings [hunk 1]
```diff
@@ -1282,11 +1282,105 @@
 /* Litewallet Partners */
 "Settings.litewallet.partners" = "Litewallet partners";
 
+<<<<<<< HEAD
```
Comment (mosadialiou) [inline loafwallet/src/Strings/Base.lproj/Localizable.strings:1285]: remaining conflict to fix

File: loafwallet/src/FlowControllers/URLController.swift [hunk 2]
```diff
@@ -64,11 +64,11 @@ class URLController : Trackable {
             return true
         case "litecoin":
             return handleBitcoinUri(url)
-        case "bitid":
-            if BRBitID.isBitIDURL(url) {
-                handleBitId(url)
-            }
-            return true
+//        case "bitid":
```
Comment (mosadialiou) [inline loafwallet/src/FlowControllers/URLController.swift:67]: we should remove this dead code

File: loafwallet/src/FlowControllers/URLController.swift [hunk 3]
```diff
@@ -103,26 +103,26 @@ class URLController : Trackable {
...
-                } else {
-                    let alert = UIAlertController(title: S.BitID.error, message: S.BitID.errorMessage, preferredStyle: .alert)
-                    alert.addAction(UIAlertAction(title: S.Button.ok, style: .default, handler: nil))
-                    self.present(alert: alert)
-                }
-            }
-        }))
-        present(alert: alert)
-    }
+//    private func handleBitId(_ url: URL) {
```
Comment (mosadialiou) [inline loafwallet/src/FlowControllers/URLController.swift:106]: we should remove this dead code

File: loafwallet/src/Extensions/UINavigationController+Extension.swift [hunk 4]
```diff
@@ -7,14 +7,19 @@
 //
 
 import UIKit
+import SwiftUI
 
 extension UINavigationController {
-
+    
+    override open func viewDidLoad() {
```
Comment (mosadialiou) [inline loafwallet/src/Extensions/UINavigationController+Extension.swift:14]: remove this overridden function

File: loafwallet/src/Constants/Strings.swift [hunk 5]
```diff
@@ -571,16 +614,17 @@ enum S {
...
 
-    enum BitID {
-        static let title = NSLocalizedString("BitID.title", value: "**BitID Authentication Request**", comment: "BitID Authentication Request alert view title")
-        static let authenticationRequest = NSLocalizedString("BitID.authenticationRequest", value: "**%1$@ is requesting authentication using your Litecoin wallet**", comment: "<sitename> is requesting authentication using your Litecoin wallet")
-        static let deny = NSLocalizedString("BitID.deny", value: "**Deny**", comment: "Deny button label")
-        static let approve = NSLocalizedString("BitID.approve", value: "**Approve**", comment: "Approve button label")
-        static let success = NSLocalizedString("BitID.success", value: "**Successfully Authenticated**", comment: "BitID success alert title")
-        static let error = NSLocalizedString("BitID.error", value: "**Authentication Error**", comment: "BitID error alert title")
-        static let errorMessage = NSLocalizedString("BitID.errorMessage", value: "**Please check with the service. You may need to try again.**", comment: "BitID error alert messaage")
+ //DEV: Removing eventually
```
Comment (mosadialiou) [inline loafwallet/src/Constants/Strings.swift:617]: we should remove this dead code

File: loafwallet/TransactionsViewController.swift [hunk 6]
```diff
@@ -405,3 +408,21 @@ class TransactionsViewController: UIViewController, UITableViewDelegate, UITable
     }
 }
  
+//struct TransactionsSwiftUIView: UIViewControllerRepresentable {
```
Comment (mosadialiou) [inline loafwallet/TransactionsViewController.swift:411]: to be removed!

File: loafwallet/TabBarViewController.swift [hunk 7]
```diff
@@ -306,10 +303,11 @@ class TabBarViewController: UIViewController, Subscriber, Trackable, UITabBarDel
...
             case 0: item.title = S.History.barItemTitle
             case 1: item.title = S.Send.barItemTitle
-            case 2: item.title = S.Receive.barItemTitle
-            case 3: item.title = S.BuyCenter.barItemTitle
+            case 2: item.title = S.LitecoinCard.barItemTitle
+            case 3: item.title = S.Receive.barItemTitle
+            case 4: item.title = S.BuyCenter.barItemTitle
             default:
-                item.title = "XXXXXX"
+                item.title = "NO-TITLE"
```
Comment (mosadialiou) [inline loafwallet/TabBarViewController.swift:310]: OR just maybe an empty title

File: loafwallet/RegistrationView.swift [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+                .previewDevice(PreviewDevice(rawValue: DeviceType.Name.iPhone8))
+                .previewDisplayName(DeviceType.Name.iPhone8)
+            
+            RegistrationView(viewModel: viewModel)
+                .previewDevice(PreviewDevice(rawValue: DeviceType.Name.iPhone12ProMax))
+                .previewDisplayName(DeviceType.Name.iPhone12ProMax)
+        }
+    }
+}
+
```
Comment (mosadialiou) [inline loafwallet/RegistrationView.swift:363]: a lot of extra lines. why?

File: loafwallet/RegistrationViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+                DispatchQueue.main.asyncAfter(deadline: .now() + 5.0) {
+                    self.isRegistering = false
+                }
+            }
+        }
+        
+       }
+     
+        if testUserID == nil {
```
Comment (mosadialiou) [inline loafwallet/RegistrationViewModel.swift:95]: do you still want to keep this `testUserID`?

File: loafwallet/Keychain.swift [hunk 10]
```diff
@@ -0,0 +1,146 @@
+//
```
Comment (mosadialiou) [inline loafwallet/Keychain.swift:1]: to be removed

File: Podfile [hunk 11]
```diff
@@ -10,6 +10,7 @@ def shared_pods
   pod 'Firebase/Crashlytics' 
   pod 'Firebase/Analytics'
   pod 'UnstoppableDomainsResolution', '~> 0.1.6'
+  pod 'KeychainAccess', '~> 4.2'
```
Comment (mosadialiou) [inline Podfile:13]: why using a 3rd party lib here. There is no Keychain API within the iOS SDK?

File: loafwallet/DataValidation.swift [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            case .password: return PasswordValidator()
+            case .mobileNumber: return MobileNumberValidator()
+            case let .requiredField(fieldName): return RequiredFieldValidator(fieldName)
+        }
+    }
+}
+ 
+struct GenericStringValidator: ValidatorConvertible {
+    func validated(_ value: String) throws -> String {
+        guard !value.isEmpty else { throw ValidationError("Must not be empty") }
```
Comment (mosadialiou) [inline loafwallet/DataValidation.swift:49]: Have to localize all the validation errors.

**Review by mosadialiou — CHANGES_REQUESTED** [review]

File: loafwallet/src/Strings/zh-Hant.lproj/Localizable.strings [hunk 13]
```diff
@@ -1275,3 +1359,4 @@
 
 /* Lookup */
 "Send.UnstoppableDomains.lookup" = "抬頭";
+>>>>>>> origin/develop
```
[the remark this answers is not in the document]
Comment (kcw-grunt) [reply loafwallet/src/Strings/zh-Hant.lproj/Localizable.strings:1362]: Thanks.

File: Podfile (hunk 11 above)
Comment (kcw-grunt) [reply Podfile:13]: It's just a wrapper to the Keychain API.  This will be refactored in v2

File: loafwallet/DataValidation.swift (hunk 12 above)
Comment (kcw-grunt) [reply loafwallet/DataValidation.swift:49]: Good catch!

File: Podfile (hunk 11 above)
Comment (mosadialiou) [reply Podfile:13]: OK sounds good! But let's avoid using a 3rd party lib to access the Keychain.
BTW can you create an issue for that.

Comment (kcw-grunt) [reply Podfile:13]: Sure...totally agreed.

Comment (kcw-grunt) [reply Podfile:13]: https://github.com/litecoin-foundation/loafwallet-ios/issues/181

File: loafwallet/RegistrationViewModel.swift (hunk 9 above)
Comment (kcw-grunt) [reply loafwallet/RegistrationViewModel.swift:95]: renamed and polished

File: loafwallet/PartnerAPIManager.swift [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+//
+
+import Foundation
+import LitewalletPartnerAPI
+
+class PartnerAPI: NSObject {
+    
+    static let shared = PartnerAPIManager()
+
+    override init() {
```
[the remark this answers is not in the document]
Comment (kcw-grunt) [reply loafwallet/PartnerAPIManager.swift:16]: it's needed by swift

File: loafwallet/LoginCardAlertView.swift [hunk 15]
```diff
@@ -70,7 +73,9 @@ struct LoginCardAlertView_Previews: PreviewProvider {
         VStack {
             Spacer()
             Text("").padding(.all, 10)
-                .loginAlertView(isShowingLoginAlert: .constant(true), message: "Login...")
+                .loginAlertView(isShowingLoginAlert: .constant(true),
+                                didFail: .constant(true),
+                                message: "Login...")
```
Comment (mosadialiou) [inline loafwallet/LoginCardAlertView.swift:78]: localize `Login...`

**Review by mosadialiou — CHANGES_REQUESTED** [review]

File: loafwallet/DataValidation.swift [hunk 16]
```diff
@@ -45,24 +41,30 @@ enum VaildatorFactory {
 }
  
 struct GenericStringValidator: ValidatorConvertible {
+    
     func validated(_ value: String) throws -> String {
-        guard !value.isEmpty else { throw ValidationError("Must not be empty") }
+        
+        guard !value.isEmpty else { throw ValidationError(S.LitecoinCard
```
Comment (mosadialiou) [inline loafwallet/DataValidation.swift:47]: You can have this in just one line for better readability. What the formatter suggests?

**Review by mosadialiou — CHANGES_REQUESTED** [review]

File: loafwallet/LoginCardAlertView.swift (hunk 15 above)
Comment (kcw-grunt) [reply loafwallet/LoginCardAlertView.swift:78]: No...that's for just a developers preview.

File: loafwallet/DataValidation.swift (hunk 16 above)
Comment (kcw-grunt) [reply loafwallet/DataValidation.swift:47]: The formatter that i'm using 'swiftformat' wraps `.` to reduce lines to 80 chars or less.  This is by design

Comment (mosadialiou) [reply loafwallet/DataValidation.swift:47]: OK

File: loafwallet/LoginCardAlertView.swift (hunk 15 above)
Comment (mosadialiou) [reply loafwallet/LoginCardAlertView.swift:78]: ok

**Review by mosadialiou — APPROVED** [review]
LGTM

**Pull request merged**